### ISSUE TITLE
Conversion to MD leaves some broken links.

### ISSUE THREAD

**Issue by rstub** [eddelbuettel/ctv-hpc#22]
As discussed in #21, the conversion script to MD leaves some broken links:
See the paragraph that I changed in the current MD file (with line breaks for better readability):

```
-   Random-number generators for parallel computing are available via the 
[rlecuyer](https://cran.r-project.org/package=rlecuyer/index.html) package, 
as well as the [sitmo](../packages/sitmo) package.
```

The first link looks odd with the /index.html at the end, while the second link is broken. Things got worse with my first commit:

```
-   Random-number generators for parallel computing are available via the 
[rlecuyer](https://cran.r-project.org/package=rlecuyer/index.html) package,
 the [rstream](../packages/rstream/index.html) package,
 the [sitmo](../packages/sitmo/index.html) package as well as
 the [dqrng](../packages/dqrng) package.
```

The first link is as it is now. All other links are broken, but it is interesting that for the last link the /index.html was removed.

This does not only affect this particular paragraph but all paragraphs with more than one link in it. In all these cases only the first link works. See for example the paragraph on Rmpi:

```
-   In recent years, the alternative MPI (Message Passing Interface) standard has become
 the de facto standard in parallel computing. It is supported in R via the 
[Rmpi](https://cran.r-project.org/package=Rmpi/index.html) by Yu. 
[Rmpi](../packages/Rmpi/index.html) package is mature yet actively maintained and offers 
access to numerous functions from the MPI API, as well as a number of R-specific extensions. 
[Rmpi](../packages/Rmpi) can be used [...]
```
The reason for this is the regular expression used to correct the links:
```
###  - make the implicit URL to packages explicit
       "-e's|../packages/\\(.*\\)/index.html|https://cran.r-project.org/package=\\1|g' ",
                         ^^^^^^^^
```

If there are more than one link in a paragraph, which is output as single line, the highlighted part of the regex will grab everything between the start of the first link and the end of the last link. This explains why the first link was transformed but with /index.html while the other links where not. Only the last link got the /index.html removed. There are ways to change the "greediness" of regexes, but I am not sure if they work with sed. The fix I am proposing is to use

```
###  - make the implicit URL to packages explicit
       "-e's|../packages/\\([^/]*\\)/index.html|https://cran.r-project.org/package=\\1|g' ",
                         ^^^^^^^^^^^
```
Instead of allowing everything between the known start and end of a link, this forbids slashes and thereby limits the regex to one link at a time.

**Comment by eddelbuettel** [comment]
We need to step back one here.  The canonical source has this:

```
    <p>
    <strong>Parallel computing: Random numbers</strong>
    </p>
    <ul>	     
      <li>Random-number generators for parallel computing are available via
        the <pkg>rlecuyer</pkg> package, as well as the <pkg>sitmo</pkg> package.
      </li>
```

No issue so far.

We convert to html via `ctv`:

```
edd@rob:~/git/ctv-hpc(master)$ r -lctv,XML -e'ctv2html(read.ctv("HighPerformanceComputing.ctv"), "/tmp/hpc.html")'
edd@rob:~/git/ctv-hpc(master)$ 
```

Still looks ok though we now have relative links:

```
      <p>
        <strong>
          Parallel computing: Random numbers
        </strong>
      </p>
      <ul>
        <li>
          Random-number generators for parallel computing are available via
        the
<a href="../packages/rlecuyer/index.html">rlecuyer</a>
          package, as well as the
<a href="../packages/sitmo/index.html">sitmo</a>
          package.
        </li>
```

We could just convert now:

```
pandoc -s -r html -w markdown_github /tmp/hpc.html -o /tmp/hpc.md
```

and then fix all URLs starting with `(../` and insert the CRAN document root.  Better?

**Comment by eddelbuettel** [comment]
It's the script, and/or I am missing something about multiple `sed` 'hits' on the same chunk.  Try this (temp) version on the ctv file as of before your patch:

```sh
#!/usr/bin/r
## if you do not have /usr/bin/r from littler, just use Rscript
##
## Copyright 2014 - 2017  Dirk Eddelbuettel
## Released under GPL-2 or later

ctv <- "HighPerformanceComputing"

ctvfile  <- paste0(ctv, ".ctv")
htmlfile <- paste0(ctv, ".html")
#mdfile   <- paste0(ctv, ".md")
mdfile   <- "newREADME.md"

## load packages
suppressMessages(library(XML))          # called by ctv
suppressMessages(library(ctv))

r <- getOption("repos")                 # set CRAN mirror
r["CRAN"] <- "https://cloud.r-project.org"
options(repos=r)

check_ctv_packages(ctvfile)             # run the check

## create html file from ctv file
ctv2html(read.ctv(ctvfile), htmlfile)



### these look atrocious, but are pretty straight forward. read them one by one
###  - call pandoc on the htmlfile, specifying html input and github-flavoured markdown
cmd <- paste0("pandoc -s -r html -w markdown_github ", htmlfile, " | ",
###  - in lines of the form  ^<a href="Word">Word.html</a>
###  - capture the 'Word' and insert it into a larger URL containing an absolute reference to task view 'Word'
              " sed -e 's|^<a href=\"\\([a-zA-Z]*\\)\\.html|<a href=\"https://cran.r-project.org/web/views/\\1.html\"|' ",
###  - deal with the header by removing extra ||, replacing |** with ** and **| with **:
              "-e's/||//g' -e's/|\\*\\*/\\*\\*/g' -e's/\\*\\*|/\\*\\* /g' -e's/|$/  /g' ",
###  - make the implicit URL to packages explicit
              "-e's|../packages/\\(.*\\)/index.html|https://cran.r-project.org/package=\\1|g' ",
              "-e's|../packages/\\(.*\\)/index.html|https://cran.r-project.org/package=\\1|g' ",
              "-e's|../packages/\\(.*\\)/index.html|https://cran.r-project.org/package=\\1|g' ",
              "-e's|../packages/\\(.*\\)|https://cran.r-project.org/package=\\1|g' ",
              "-e's|../packages/\\(.*\\)|https://cran.r-project.org/package=\\1|g' ",
              "-e's|../packages/\\(.*\\)|https://cran.r-project.org/package=\\1|g' ",
              "-e's|../packages/\\(.*\\)|https://cran.r-project.org/package=\\1|g' ",
              "-e's|../packages/\\(.*\\)|https://cran.r-project.org/package=\\1|g' ",
              "-e's|../packages/\\(.*\\)|https://cran.r-project.org/package=\\1|g' ",
              "-e's|../packages/\\(.*\\)|https://cran.r-project.org/package=\\1|g' ",
              "-e's|../packages/\\(.*\\)|https://cran.r-project.org/package=\\1|g' ",
              "-e's|../packages/\\(.*\\)|https://cran.r-project.org/package=\\1|g' ",
              "-e's|../packages/\\(.*\\)|https://cran.r-project.org/package=\\1|g' ",
              "-e's|../packages/\\(.*\\)|https://cran.r-project.org/package=\\1|g' ",
###  - write out mdfile
              "> ", mdfile)
print(cmd)
system(cmd)                             # run the conversion

#unlink(htmlfile)                        # remove temporary html file

cat("Done.\n")
```

**Comment by rstub** [comment]
I am away from my computer so I am not sure what you are trying to show.
However, the issue is not about the regex  only matching the first link on
the line. Instead the regex matches all links at once. I will try to make a
better example later.

Dirk Eddelbuettel <notifications@github.com> schrieb am So. 10. Juni 2018
um 02:00:

**Comment by eddelbuettel** [comment]
This is not urgent. I just showed you a 'in-progress' update to the script which addresses all your concerns.  No more remaining `...` in the resulting markdown.  And, as best as I can tell, it has _nothing_ to do with line beginnings. It is simply an `sed` limitation that it seems to finish at the first hit per line which is why I repeated the expressions above.  There is most certainly a better way -- but I've been busy too.  We'll get to it next week.

**Comment by rstub** [comment]
The resulting md file looks fine with the exception of the first link on a line with multiple links, e.g.:
```
[...] R via the [Rmpi](https://cran.r-project.org/package=Rmpi/index.html) by Yu. 
[Rmpi](https://cran.r-project.org/package=Rmpi) package is mature [...]
```

That trailing `/index.html` is superfluous, though the link works. However, I think the patch is unnecessarily complex, since `sed` is capable of processing multiple hits on one line. The `g` at the end of the pattern takes care of that. Here the patch that was part of the commit I removed from #21:
```
$ git diff ctv2md.r
diff --git a/ctv2md.r b/ctv2md.r
index dfb1019..76c644f 100755
--- a/ctv2md.r
+++ b/ctv2md.r
@@ -35,7 +35,7 @@ cmd <- paste0("cat ", htmlfile,
 ###  - deal with the header by removing extra ||, replacing |** with ** and **| with **:
               "sed -e's/||//g' -e's/|\\*\\*/\\*\\*/g' -e's/\\*\\*|/\\*\\* /g' -e's/|$/  /g' ",
 ###  - make the implicit URL to packages explicit
-              "-e's|../packages/\\(.*\\)/index.html|https://cran.r-project.org/package=\\1|g' ",
+              "-e's|../packages/\\([^/]*\\)/index.html|https://cran.r-project.org/package=\\1|g' ",
 ###  - write out mdfile
               "> ", mdfile)
```
This fixes all links without any trailing `/index.html`.

**Comment by eddelbuettel** [comment]
The `g` should -- but doesn't. 

I played with that for several minutes when I made the change.  Something is up there.

**Comment by rstub** [comment]
**tldr**: The `g` works as expected with my patch. Please give it a try.

Here some examples, always with the same line containing two links to packages.

Original regex:
```
$ echo "one ../packages/foo/index.html two ../packages/bar/index.html three" | \
       sed -e's|../packages/\(.*\)/index.html|https://cran.r-project.org/package=\1|g'
one https://cran.r-project.org/package=foo/index.html two ../packages/bar three
```
The first link is working but has a trailing `/index.html`. The second link has the `/index.html` removed, but is otherwise broken. Explanation: the `.*` matches `foo/index.html two ../packages/bar`.

Patched regex without `g`:
```
$ echo "one ../packages/foo/index.html two ../packages/bar/index.html three" | \
       sed -e's|../packages/\([^/]*\)/index.html|https://cran.r-project.org/package=\1|'
one https://cran.r-project.org/package=foo two ../packages/bar/index.html three
```
Now the first link es correct, but the second link is unchanged. Explanation: the `[^/]*` matches `foo`.

Patched regex with `g`:
```
$ echo "one ../packages/foo/index.html two ../packages/bar/index.html three" | \
       sed -e's|../packages/\([^/]*\)/index.html|https://cran.r-project.org/package=\1|g'
one https://cran.r-project.org/package=foo two https://cran.r-project.org/package=bar three
```

Now both links are correct. Explanation: the `[^/]*` matches `foo` and `bar`.

**Comment by eddelbuettel** [comment]
Arguing what happens on the shell is one thing,  it is a little different still in the script.  But your point stands -- my regexp was not ideal, and your version was better.  It still needed one more change.  

Better version committed in 888abe1e332d4a8285d9c046861548d544be1224.
